fix recursion and other minor fixes

ayushnawal opened the pull request:
Fixes #18 

Tested

removed `save_pattern` , `g_init` etc. as these functions were not defined 

@quozl , @chimosky

ayushnawal committed:
Remove recursion on activity resize

ayushnawal committed:
Fix NameError for 'GRID_CELL_SIZE'

ayushnawal committed:
remove undefined function calls

ayushnawal committed:
return on pygame.QUIT

ayushnawal commented:
@srevinsaju please review

srevinsaju commented:
Tested, looks good except occassional lag on pressing the button. I am not smart enough to complete the game, it looks trivial. There is a flake warning and error on each and every line.
I guess this PR is already an improvement over the current status..
Thanks @ayushnawal

quozl commented:
As the recursion was driven by display size changes, it is critical to test under that scenario.

Tested on OLPC XO-4 'cause it was the easiest to trigger display size changes on; but the activity used only a tiny area of the screen.  I traced this to a regression introduced in ec6626557807291926155082b3b8bd48d8bf6187, but I don't know the cause.

Tested on OLPC NL3 with Ubuntu 20.04 and Sugar 0.116.  The NL3 is 1366x768, and an attached HDMI monitor supported up to 1920x1080.  The activity displayed correctly, but changing the display size between 1366x768 and 1920x1080 gave a black screen with nothing unusual in logs.

Tested on QEMU VM with Ubuntu 20.04 and Sugar 0.116.  The activity displayed correctly at the resolution in use, but changing the display size didn't change the activity layout.  The activity continue at the first detected display size.

Summary: your change lost the feature of adapting to display size.

You didn't mention how you tested for display size changes.  Here is my test method;

* install _screen_,
* open _Terminal_ and start a _screen_ session,
* use _ssh_ to log in to the virtual machine and `su - user` where _user_ is the username in which Sugar is logged in, then `screen -x` to share the _screen_ session,
* start the Across and Down activity,
* use the _ssh_ connection to change the display size;

```
xrandr --output Virtual-0 --mode 1024x768
xrandr --output Virtual-0 --mode 1366x768
```

Hope that helps with your testing.

ayushnawal commented:
Thanks @quozl 

Sorry, I was slightly busy past week due to my college exams so I didn't look at your review comments.

This is how I tested my changes before:
* run activity
* using display settings on ubuntu, changed to other display options.
* no error, warnings, traceback. 
* changed back to normal display settings.
* again no error.

It was working as expected on my system.

Tested on ubuntu 18.04 VM

quozl commented:
Ah, that's interesting, thanks.  Display settings in Ubuntu is not available when running Sugar.  Are you testing outside Sugar?

ayushnawal commented:
> Are you testing outside Sugar?

Yes :sweat_smile:, tested outside sugar using `sugar-activity3 .`, as. I wanted to change the display settings.

I will test it with your testing method. thanks for that as I was unaware of a method to change the display using terminal only. 

I will update you with the test results.

JuiP commented:
@quozl why was this change ec66265 necessary? Doesn't [this](https://github.com/sugarlabs/across-and-down-activity/blob/2459de3c7ed587da12a90f3ac2173fc14c07e573/sugargame/event.py#L100) do the same thing?
@pro-panda

srevinsaju commented:
> why was this change ec66265 necessary? Doesn't this do the same thing?

@JuiP I am afraid, it does not

* `pygame.event.Event()` is an object created in `sugargame.event`. 
* That `event.Event` [here](https://github.com/sugarlabs/across-and-down-activity/blob/2459de3c7ed587da12a90f3ac2173fc14c07e573/sugargame/event.py#L100) is a subclass of `pygame.VIDEORESIZE`. It is emitted by sugargame when `configure-event` signal is emitted

https://github.com/sugarlabs/across-and-down-activity/commit/ec6626557807291926155082b3b8bd48d8bf6187 checks for the event signal, and performs a certain function when the event is created. A new window is created, or when display size is changed (as I understand)

This is how it happens
* sugargame.event receives signal `configure-event`, posts `pygame.event.Event` which is `VIDEORESIZE` to the `pygame.event` poll
* the activity, receives the `pygame.event.Event` when `pygame.event.get()` is called
* The `for` loop iterates through each event, and finds [this](https://github.com/sugarlabs/across-and-down-activity/blob/2459de3c7ed587da12a90f3ac2173fc14c07e573/sugargame/event.py#L100) event, and then processes what is supposed to be done in https://github.com/sugarlabs/across-and-down-activity/commit/ec6626557807291926155082b3b8bd48d8bf6187

This has a very close relation to how signals and slots work in `PyQt` if you do know about `pyqt` this might help you to understand better. :smile:

Even I had confusions about `sugargame`, but porting some activities to `sugargame v1.3` helped me understand how `sugargame` works. `sugargame` is a warapper `Gtk.VBox` which allows pygame to be run within it. So `pygame` may not receive all the signals which it was likely to receive when its being run as a Application, i.e with its own window and frame (rather than within a sugar window). So afaik, sugargame converts the signals in a Sugar GTK window to `pygame`.

quozl commented:
Pygame is most frequently used with fullscreen programs.  Sugar adds a toolbar.  So `_resize_cb` translates the GTK resize event into a synthesised Pygame event.  Once the Pygame event loop releases the event (in the correct order), the `set_mode` changes the Pygame window size.
